File: synthesisDataGen.m
m=500;
C=1;
X=10000*rand(64,10000);
d=size(X,1);
k=3; %k is the rank
%u=(-1*ones(d,k)+2*rand(d,k))/10;
%v=(-1*ones(k,m)+2*rand(k,m))/10;
u=randn(d,k);
v=randn(k,m);
realU=u;
realV=v;
realpara=realU*realV;
train={};
for i=1:500
    realw=realpara(:,i);
    rating=realw'*X;
    sampleindex=randi(10000,50,1);
    usercol=ones(50,1)*i;
    samplecol=sampleindex;
    ratecol=rating(sampleindex)';
    train{i}=[usercol,samplecol,ratecol];
end
train=train';
testset=genTestset(train,5);        % 5 per user as the testset
testset= arrayfun(@(x) testset(testset(:,1)==x,:),unique(testset(:,1)),'uniformoutput',false);
syngenpair;
ranksvmgrad;

File: julia_version/synthesisdatagen.jl
function syngentrain(m=1000,mo=10000,pir=40,d=64,k=20)
	# m=100 # number of users
	# C=1  # constant control the loss function
	# mo=100 # number of movies
	# pir=15 # number of pairs
	srand(49)
	X=randn(d,mo) # d=64 movies=100
	# k=5  # rank=5
	#d=size(X,1)
	srand(50)
	u=randn(d,k)
	srand(51)
	v=randn(k,m)     #get the real value of u,v
	realpara=u*v
	train=zeros(1,3)
	for i=1:m
    		realw=realpara[:,i]
    		rating=X'*realw
		randp=randperm(mo) # generate nonrep
		randp=randp'' 
		samplemovies=randp[1:pir,1]
    		usercol=ones(Int64,pir,1)*i;  # each user has 15 ratings
    		samplecol=samplemovies
    		ratecol=rating[samplemovies]
    		temp=[usercol samplecol ratecol]
    		train=[train;temp]
	end
	train=train[2:end,:]
	return train,X
end
